### PR TITLE
Refactor uniffi-runtime-napi into layers

### PR DESCRIPTION
**Pull request by jhugman**
…to add unload lifecycle methods
…to memoize dynamic FFI lookups
…to unlock the potential of a JSI player.

The PR is large but is really just moving existing code around: no new functionality is added, and our test coverage is good.

### CODE REVIEW

File: runtimes/napi/tests/unload.test.mjs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+});
+
+test("re-register after unload works", () => {
+  const { nativeModule: nm1 } = openAndRegister();
+  nm1.unload();
+  const { nativeModule: nm2, bindings: b2 } = openAndRegister();
+  const status = { code: 0 };
+  assert.strictEqual(b2.uniffi_test_fn_add(4, 5, status), 9);
+  nm2.unload();
+});
```
Comment (1egoman) [inline runtimes/napi/tests/unload.test.mjs:58]: What is expected to happen if you were to call `nativeModule.register` twice in a row? Like:
```
const nativeModule = UniffiNativeModule.open(LIB_PATH);
const bindingsA = nativeModule.register({
  symbols: SYMBOLS,
  structs: {},
  callbacks: {},
  functions: FUNCTIONS,
});
const bindingsB = nativeModule.register({
  symbols: SYMBOLS,
  structs: {},
  callbacks: {},
  functions: FUNCTIONS,
});
```

This might be another edge case to test in here.

**Review by 1egoman — APPROVED** [review]
Scanned through it all, generally made sense.

**Comment by jhugman** [comment]
Closed #379
